bsip-0061.md

```diff
@@ -0,0 +1,169 @@
+    - *account_to_unlock* should have sufficient balances to perform this operation.
+    - Check if *account_to_unlock* is locked in unlock_account_evaluator.
+    - Check if *account_to_unlock* is authorized to perform the operation (transaction signed with previous authority).
+    - Undo "lock"-operation. See Discussion - Unlocking account.
+
+# Specifications
+## Prevent circular dependencies
+```
+account_create_evaluator::do_evaluate( )
+{  ...
+   if( head_block_time() < HARDFORK_PREVENT_CYCLES_TIME )
+   {
+       graphene::chain::verify_cycled_authority( account );
+   }
+   ...
+}
+
+account_update_evaluator::do_evaluate( )
+{  ...
+   if( head_block_time() < HARDFORK_PREVENT_CYCLES_TIME )
+   {
+       graphene::chain::verify_cycled_authority( account );
+   }
+   ...
+}
+```
+```
+verify_cycled_authority( account_id )
+{
+   sign_state state( initial_parameters, verify_only=true );
+
+   GRAPHENE_ASSERT( state.check_authority(account_id), "Missing Authority" );
+}
+```
+```
+class sign_state
+{
+    sign_state( old_parameters, bool verify_only = false )
+
+    //
+    // New method added to parametrize class behaviour
+    //
+    bool is_key_available(const public_key_type &key)
+    {
+        if (verify_only)
+        {
+           return true
+        }
+
+        auto pk = available_keys.find(key);
+        return (pk != available_keys.end());
+    }
+
+    ...
+    bool verify_only
+}
+```
+
+# Discussion
+## Unlocking account
+Operation that prior to locked state will be fully re-done. There is an option - only owner or active authorities can be restored.
+
+# Examples
+## Unlocking account using previous authority. Fully re-done operation
+There is following use case:
+
+    1) Account "Alice" was created
+    2) Account "Alice" was updated with new authorities and options (opts 1) - here: prior state
+    3) Account "Alice" was updated with new options (opts 2)
+    4) Account "Alice" was updated with new options (opts 3)
+    5) Account "Alice" was updated with new authorities - here: locked state
+
+To unlock "Alice" we need to reapply update operation at step 2. But latest account options (opts 3) at step 4 will be lost.
+
+## Unable to unlock account
+Use case:
+
+    1) Account "Alice" delegates its authority to "Bob" - here: prior state
+    2) Account "Charlie" delegates its authority to "Alice"
+    3) Account "Alice" delegates its authority to "Charlie" (transaction was signed with Bob's private key) - here: locked state
+    4) Account "Bob" delegates its authority to "Charlie" - here: locked state
+
+To unlock "Alice" we need to reapply update operation at step 1. But "Bob" also locked, we have double lock here.
+
+# Summary for Shareholders
+We offer to develop a solution for BitShares users.
+This solution allows unlocking accounts, which are locked by mistake.
+The most important goal is to return access to the wallets and unlock their money.
+
```

Comment:
Please add a cost/benefit discussion.
How many such locked accounts exist today, how much are the locked balances worth, and what would it cost to implement this proposal?

Reply:
Updated.

Reply:
cost/benefit discussion?

Reply:
Could you please clarify the **_cost/benefit_**?

Reply:
> How many such locked accounts exist today, how much are the locked balances worth, and what would it cost to implement this proposal?

Reply:
Today exist over 300 locked accounts. 
The most important goal is to prevent irreversible blocking of accounts and help to regain access to the already blocked accounts and their balances.

Reply:
Thanks. Didn't expect that many...

Reply:
Is there a list of these accounts somewhere?

Reply:
List of locked accounts
1.2.352708
1.2.363525
1.2.312657
1.2.133051
1.2.100886
1.2.117772
1.2.182288
1.2.1150329
1.2.380405
1.2.777808
1.2.863848
1.2.996565
1.2.989841
1.2.31223
1.2.166289
1.2.169641
1.2.186918
1.2.206921
1.2.166880
1.2.12418
1.2.98043
1.2.501729
1.2.740449
1.2.665269
1.2.960390
1.2.1259335
1.2.179504
1.2.155110
1.2.262650
1.2.380718
1.2.378612
1.2.456139
1.2.403919
1.2.107093
1.2.165413
1.2.165714
1.2.184024
1.2.907383
1.2.843615
1.2.834573
1.2.1057592
1.2.1218022
1.2.405903
1.2.265235
1.2.355200
1.2.433810
1.2.460835
1.2.122787
1.2.93429
1.2.665231
1.2.443082
1.2.1107208
1.2.1147487
1.2.1150333
1.2.290543
1.2.151149
1.2.163782
1.2.515217
1.2.96015
1.2.20080
1.2.182258
1.2.417925
1.2.663843
1.2.608383
1.2.1150281

Reply:
These are 65 accounts, but not about 300?

Reply:
We decided to change the search algorithm according to the comments - https://github.com/bitshares/bsips/issues/94

As a result, the number of locked accounts has decreased.

List of the locked balances - https://docs.google.com/document/d/1Dmcr9QzSWnCbSKcpaN08iENJyn6AWogbKjZ_iHH4OM0/edit

Reply:
We have summarised the most liquid currencies of locked accounts.

Here the list:
BTS - 1,641,712.7494
USD - 14,546.7176
BTC - 6.86072047
CNY - 16,396.3406
ETH - 0.278826
LTC - 8.60401198
DASH - 0.23722333
ICOO - 348.1812
OBITS - 3,102.1454

Total locked budget is 144,619.841 bitUSD

As you can see the total budget is huge. We offer to unlock these accounts. 

Reply:
Thanks. That is indeed significant.

How about taking 10% from each account as payment for the implementation? (Seriously. The accounts are locked which means a 100% loss for the owner. We're offering to return 90%.)

Reply:
> How about taking 10% from each account as payment for the implementation? (Seriously. The accounts are locked which means a 100% loss for the owner. We're offering to return 90%.)

It makes sense.
@pmconrad  could you please share your idea of how this can be implemented in a technical way?

Reply:
At unlock time, walk through the account balances and auto-transfer 10% from each non-zero balance to a specific account, e. g. committee-account. That's pretty straightforward.

The non-technical side may be more tricky. :-/